controller/lib/hal/uart_dma.cpp

```diff
@@ -0,0 +1,256 @@
+// CPU is notified via interrupt.
+
+// This driver also provides Character Match interrupt on reception.
+// UART will issue an interrupt upot the reception of the specified
+// character.
+
+extern UART_DMA dmaUART;
+
+// Performs UART3 initialization
+void UART_DMA::init(int baud) {
+  // Set baud rate register
+  uart->baud = CPU_FREQ / baud;
+
+  uart->ctrl3.s.dmar = 1;         // set DMAR bit to enable DMA for receiver
+  uart->ctrl3.s.dmat = 1;         // set DMAT bit to enable DMA for transmitter
+  uart->ctrl3.s.ddre = 1;         // DMA is disabled following a reception error
+  uart->ctrl2.s.rtoen = 1;        // Enable receive timeout feature
+  uart->ctrl2.s.addr = matchChar; // set match char
+
+  uart->ctrl3.s.eie = 1; // enable interrupt on error
+
+  uart->request |= (1 << 3); // Clear RXNE flag before clearing other flags
+  uart->intClear = (1 << 11) | (1 << 3) | (1 << 1); // Clear error flags
+
+  uart->ctrl1.s.te = 1; // enable transmitter
+  uart->ctrl1.s.re = 1; // Enable receiver
+  uart->ctrl1.s.ue = 1; // enable uart
+
+  // TODO enable parity checking?
+
+  dma->channel[rxCh].config.priority = 0b11; // high priority
+  dma->channel[rxCh].config.teie = 1;        // interrupt on error
+  dma->channel[rxCh].config.htie = 0;        // no half-transfer interrupt
+  dma->channel[rxCh].config.tcie = 1;        // interrupt on DMA complete
+
+  dma->channel[rxCh].config.mem2mem = 0;  // memory-to-memory mode disabled
+  dma->channel[rxCh].config.msize = 0b00; // 8 bits
+  dma->channel[rxCh].config.psize = 0b00; // 8 bits
+  dma->channel[rxCh].config.memInc = 1;   // increment destination (memory)
+  dma->channel[rxCh].config.perInc = 0;   // don't increment source
+                                          // (peripheral) address
+  dma->channel[rxCh].config.circular = 0; // not circular
+  dma->channel[rxCh].config.dir = 0;      // peripheral to memory
+
+  dma->channel[txCh].config.priority = 0b11; // high priority
+  dma->channel[txCh].config.teie = 1;        // interrupt on error
+  dma->channel[txCh].config.htie = 0;        // no half-transfer interrupt
+  dma->channel[txCh].config.tcie = 1;        // DMA complete interrupt enabled
+
+  dma->channel[txCh].config.mem2mem = 0;  // memory-to-memory mode disabled
+  dma->channel[txCh].config.msize = 0;    // 8 bits
+  dma->channel[txCh].config.psize = 0;    // 8 bits
+  dma->channel[txCh].config.memInc = 1;   // increment source (memory) address
+  dma->channel[txCh].config.perInc = 0;   // don't increment dest (peripheral)
+                                          // address
+  dma->channel[txCh].config.circular = 0; // not circular
+  dma->channel[txCh].config.dir = 1;      // memory to peripheral
+}
+
+// Sets up an interrupt on matching char incomming form UART3
+void UART_DMA::charMatchEnable() {
+  uart->intClear |= (1 << 17); // Clear char match flag
+  uart->ctrl1.s.cmie = 1;      // Enable character match interrupt
```

Comment:
It would be better to just set the bit in the intClear register rather then perform the read/modify/write.
I think the intClear register always reads zero although the manual isn't really clear on that.  If it does, then just setting the bit is the same as the read/modify/write.  If intClear can return a non-zero value, then this code will clear more bits then are intended.

Reply:
Good idea, fixed also for `uart->request` register